sklearn/metrics/_pairwise_distances_reduction/_radius_neighbors.pyx.tp

```diff
@@ -503,27 +488,35 @@ cdef class EuclideanRadiusNeighbors{{name_suffix}}(RadiusNeighbors{{name_suffix}
     ) nogil:
         cdef:
             ITYPE_t i, j
-            DTYPE_t squared_dist_i_j
-            ITYPE_t n_X = X_end - X_start
-            ITYPE_t n_Y = Y_end - Y_start
-            DTYPE_t *dist_middle_terms = self.gemm_term_computer._compute_distances_on_chunks(
+            ITYPE_t pair_index = 0
+            DTYPE_t sqeucliean_dist_i_j
+
+            DTYPE_t *dist_middle_terms = self.gemm_term_computer._compute_dist_middle_terms(
                 X_start, X_end, Y_start, Y_end, thread_num
             )
 
         # Pushing the distance and their associated indices in vectors.
-        for i in range(n_X):
-            for j in range(n_Y):
+        for i in range(X_start, X_end):
+            for j in range(Y_start, Y_end):
                 # Using the squared euclidean distance as the rank-preserving distance:
                 #
                 #             ||X_c_i||² - 2 X_c_i.Y_c_j^T + ||Y_c_j||²
                 #
-                squared_dist_i_j = (
-                    self.X_norm_squared[i + X_start]
-                    + dist_middle_terms[i * n_Y + j]
-                    + self.Y_norm_squared[j + Y_start]
+                sqeucliean_dist_i_j = (
+                    self.X_norm_squared[i]
+                    + dist_middle_terms[pair_index]
+                    + self.Y_norm_squared[j]
                 )
-                if squared_dist_i_j <= self.r_radius:
-                    deref(self.neigh_distances_chunks[thread_num])[i + X_start].push_back(squared_dist_i_j)
-                    deref(self.neigh_indices_chunks[thread_num])[i + X_start].push_back(j + Y_start)
 
+                # Guard against eventual -0. and NaN caused by catastrophic
+                # cancellation (e.g. if X is Y)
+                sqeucliean_dist_i_j = (
+                    sqeucliean_dist_i_j if 0. < sqeucliean_dist_i_j == sqeucliean_dist_i_j else 0.
+                )
```

Comment:
I find the ternary operator not too readable here. Also I don't understand the `sqeucliean_dist_i_j == sqeucliean_dist_i_j` part of the condition.

```suggestion
                if sqeucliean_dist_i_j < 0.:
                    sqeucliean_dist_i_j = 0.
```

Reply:
`sqeucliean_dist_i_j == sqeucliean_dist_i_j` is False if `sqeucliean_dist_i_j` is `NaN`.

The ternary operator is convenient, and I think the comment can better phrased and placed closer to this operator. What do you think?

Alternatively, I am OK introducing the conditional branch you suggest, which (replacing comments) could be:

```python
                # < Remove previous comment> 
                if (
                    # <insert negative zero guard comment>
                    sqeucliean_dist_i_j < 0. or
                    # < explain the NaN guard bellow>
                    sqeucliean_dist_i_j != sqeucliean_dist_i_j
                ):
                    sqeucliean_dist_i_j = 0.
```

Reply:
192dd92eb2280c66f081a249d61b2822fa96f0a5 introduced a proposal for a new comment.